### PR TITLE
Produce errors on ref-like in arguments that may compromise stack safety rules

### LINKED ISSUE #27357: Incorrect and inconsistent error for mismatching byref arguments' escape scopes
**Version Used**: 

Microsoft Visual Studio Professional 2017 
Version 15.7.2
VisualStudio.15.Release/15.7.2+27703.2018
Microsoft .NET Framework
Version 4.7.03056

**Steps to Reproduce**:

```cs
using System;

static class C
{
  static void M(Span<int> s1)
  {
    Span<int> s2 = stackalloc int[10];
    M(s2, s1); // no errors here
    M(s2, in s1); // errors
  }

  public static void M(Span<int> x, in Span<int> y)
  {
  }
}
```

**Expected Behavior**:
Either errors for both invocations or for neither of them

**Actual Behavior**:
Errors reported only for the second invocation.
```
Error CS8350: This combination of arguments to 'C.M(Span<int>, in Span<int>)' is disallowed because it may expose variables referenced by parameter 'x' outside of their declaration scope
Error CS8352: Cannot use local 's2' in this context because it may expose referenced variables outside of their declaration scope
```

**Notes**
These errors should be reported either for both invocations or for neither of them since these calls are equivalent.
I personally think that this error should not be reported for `in` arguments at all because possible byvalue leaks to `ref`/`out` arguments are already reported regardless of `in` arguments and their byref assignments will be limited to the callee.

### PR DESCRIPTION
**Pull request by OmarTawfik**
Fixes #27357

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs [hunk 1]
```diff
@@ -1282,8 +1281,8 @@ bool isRefEscape
                         break;
                     }
 
-                    var refKind = argRefKindsOpt.IsDefault ? RefKind.None : argRefKindsOpt[argIndex];
-                    if (refKind != RefKind.None && argument.Type?.IsByRefLikeType == true)
+                    var paramIndex = argsToParamsOpt.IsDefault ? argIndex : argsToParamsOpt[argIndex];
+                    if (parameters[paramIndex].RefKind != RefKind.None && argument.Type?.IsByRefLikeType == true)
```
Comment (OmarTawfik) [inline src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs]: Check `GetInvocationEscapeScope()` and `GetEffectiveRefKindAndMarkMatchedInParameter()`

File: src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs [hunk 2]
```diff
@@ -1298,7 +1297,7 @@ bool isRefEscape
                     {
                         var argument = argListArgs[argIndex];
                         var refKind = argListRefKindsOpt.IsDefault ? RefKind.None : argListRefKindsOpt[argIndex];
-                        if (refKind != RefKind.None && argument.Type?.IsByRefLikeType == true)
```
Comment (OmarTawfik) [inline src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs]: Add a test here.

**Comment by OmarTawfik** [comment]
@dotnet/roslyn-compiler for review.

File: src/Compilers/CSharp/Portable/Symbols/RefKindExtensions.cs [hunk 3]
```diff
@@ -30,5 +30,20 @@ public static RefKind GetRefKind(this SyntaxKind syntaxKind)
...
+        public static bool IsWritableReference(this RefKind refKind)
+        {
+            switch (refKind)
+            {
+                case RefKind.Ref:
+                case RefKind.Out:
+                    return true;
+                case RefKind.None:
+                case RefKind.In:
+                    return false;
```
Comment (jaredpar) [inline src/Compilers/CSharp/Portable/Symbols/RefKindExtensions.cs]: Why aren't we checking RefReadOnly here?

Comment (OmarTawfik) [reply src/Compilers/CSharp/Portable/Symbols/RefKindExtensions.cs]: It is the same value as `RefKind.In`.

File: src/Compilers/CSharp/Test/Semantic/Semantics/SpanStackSafetyTests.cs [hunk 4]
```diff
@@ -1440,5 +1440,208 @@ static void Main()
...
+        M1(s1);
+    }
+    
+    static void M1(Span<int> s1)
+    {
+		var obj = new C();
+		Span<int> s2 = stackalloc int[2];
+
+        M2(ref s1, out s2);         // one
+        M2(ref s2, out s1);         // two
```
Comment (agocke) [inline src/Compilers/CSharp/Test/Semantic/Semantics/SpanStackSafetyTests.cs]: Does `M2(ref s2, out s2)` work? What about `M2(ref s1, out s1)`?

Comment (OmarTawfik) [reply src/Compilers/CSharp/Test/Semantic/Semantics/SpanStackSafetyTests.cs]: There should be no errors for that. I'll modify the test.

---
In reply to: [197882267](https://github.com/dotnet/roslyn/pull/27883#discussion_r197882267) [](ancestors = 197882267)

**Comment by OmarTawfik** [comment]
@agocke any other comments?
@jaredpar for merge approval

**Pull request merged**